File: flink-code-java-0421/src/main/java/com/atguigu/day05/IntervalJoinExample.java
package com.atguigu.day05;

import org.apache.flink.api.common.eventtime.SerializableTimestampAssigner;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.java.tuple.Tuple3;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.KeyedStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.co.ProcessJoinFunction;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.util.Collector;

public class IntervalJoinExample {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.setParallelism(1);

        KeyedStream<Tuple3<String, Long, String>, String> clickStream = env
                .fromElements(
                        Tuple3.of("user_1", 10 * 60 * 1000L, "click"),
                        Tuple3.of("user_1", 16 * 60 * 1000L, "click"),
                        Tuple3.of("user_2", 20 * 60 * 1000L, "click")
                )
                .assignTimestampsAndWatermarks(
                        WatermarkStrategy.<Tuple3<String, Long, String>>forMonotonousTimestamps()
                                .withTimestampAssigner(new SerializableTimestampAssigner<Tuple3<String, Long, String>>() {
                                    @Override
                                    public long extractTimestamp(Tuple3<String, Long, String> click, long l) {
                                        return click.f1;
                                    }
                                })
                )
                .keyBy(r -> r.f0);

        KeyedStream<Tuple3<String, Long, String>, String> browseStream = env
                .fromElements(
                        Tuple3.of("user_1", 5 * 60 * 1000L, "browse"),
                        Tuple3.of("user_1", 6 * 60 * 1000L, "browse")
                )
                .assignTimestampsAndWatermarks(
                        WatermarkStrategy.<Tuple3<String, Long, String>>forMonotonousTimestamps()
                                .withTimestampAssigner(new SerializableTimestampAssigner<Tuple3<String, Long, String>>() {
                                    @Override
                                    public long extractTimestamp(Tuple3<String, Long, String> browse, long l) {
                                        return browse.f1;
                                    }
                                })
                )
                .keyBy(r -> r.f0);

        clickStream
                .intervalJoin(browseStream)
                .between(Time.minutes(-10), Time.minutes(0))
                .process(new ProcessJoinFunction<Tuple3<String, Long, String>, Tuple3<String, Long, String>, String>() {
                    @Override
                    public void processElement(Tuple3<String, Long, String> click, Tuple3<String, Long, String> browse, Context context, Collector<String> collector) throws Exception {
                        collector.collect(click + " => " + browse);
                    }
                })
                .print();

        env.execute();
    }
}

File: Flink0105/src/main/scala/com/atguigu/day5/IntervalJoinExample.scala
package com.atguigu.day5

import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.co.ProcessJoinFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.util.Collector

object IntervalJoinExample {
  def main(args: Array[String]): Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    // 基于间隔的join只能使用事件时间
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    env.setParallelism(1)

    // 点击流
    val clickStream = env
      .fromElements(
        ("1", "click", 3600 * 1000L)
      )
      .assignAscendingTimestamps(_._3)
      .keyBy(_._1)

    // 浏览流
    val browseStream = env
      .fromElements(
        ("1", "browse", 2000 * 1000L),
        ("1", "browse", 3100 * 1000L),
        ("1", "browse", 3200 * 1000L),
        ("1", "browse", 4000 * 1000L),
        ("1", "browse", 7200 * 1000L)
      )
      .assignAscendingTimestamps(_._3)
      .keyBy(_._1)

    clickStream
      .intervalJoin(browseStream)
      // 3600s interval join (3000s ~ 4100s)
      .between(Time.seconds(-600), Time.seconds(500))
      .process(new MyIntervalJoin)
      .print()

    // 和上面的写法等价，需要将区间从(-600, 500)改成(-500, 600)
    // a.ts - 600 < b.ts < a.ts + 500
    // a.ts < b.ts + 600
    // a.ts > b.ts - 500
    // b.ts - 500 < a.ts < b.ts + 600
    browseStream
        .intervalJoin(clickStream)
        .between(Time.seconds(-500), Time.seconds(600))
        .process(new MyIntervalJoin)
        .print()

    env.execute()
  }

  class MyIntervalJoin extends ProcessJoinFunction[(String, String, Long), (String, String, Long), String] {
    override def processElement(left: (String, String, Long), right: (String, String, Long), ctx: ProcessJoinFunction[(String, String, Long), (String, String, Long), String]#Context, out: Collector[String]): Unit = {
      // left 来自第一条流； right 来自第二条流
      out.collect(left + " =====> " + right)
    }
  }
}